File: sdk/javascript/examples/auto-heartbeat.ts
/**
 * Example: Auto Heartbeat
 *
 * This example demonstrates how to maintain a regular heartbeat
 * with the EKET server to keep the agent's status alive.
 *
 * Run: npx ts-node sdk/javascript/examples/auto-heartbeat.ts
 */

import { EketClient } from '../src/index.js';

const SERVER_URL = process.env.EKET_SERVER_URL ?? 'http://localhost:8080';
const HEARTBEAT_INTERVAL_MS = 30_000; // 30 seconds
const RUN_DURATION_MS = 120_000;      // 2 minutes

async function main() {
  const client = new EketClient({ serverUrl: SERVER_URL });

  // Step 1: Register agent
  console.log('Registering agent...');
  const { instance_id, heartbeat_interval } = await client.registerAgent({
    agent_type: 'custom',
    role: 'slaver',
    specialty: 'backend',
    capabilities: ['typescript', 'node'],
  });
  console.log(`✓ Registered: ${instance_id}`);
  console.log(`  Server-recommended heartbeat interval: ${heartbeat_interval}s`);

  // Step 2: Send heartbeats at regular intervals
  console.log(`\nStarting heartbeat loop (every ${HEARTBEAT_INTERVAL_MS / 1000}s)...`);

  let beatCount = 0;

  const interval = setInterval(async () => {
    try {
      beatCount += 1;
      const result = await client.sendHeartbeat(instance_id, {
        status: 'idle',
        current_task: undefined,
        progress: 0,
      });

      console.log(`[${new Date().toISOString()}] ✓ Heartbeat #${beatCount} sent`);
      console.log(`  Server time: ${result.server_time}`);

      // Check for pending messages
      if (result.messages.length > 0) {
        console.log(`  📧 ${result.messages.length} pending message(s):`);
        for (const msg of result.messages) {
          console.log(`    - [${msg.type}] from ${msg.from}`);
        }
      }
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : String(err);
      console.error(`[${new Date().toISOString()}] ✗ Heartbeat failed: ${message}`);
    }
  }, HEARTBEAT_INTERVAL_MS);

  // Step 3: Stop after RUN_DURATION_MS
  await new Promise<void>((resolve) => {
    setTimeout(() => {
      clearInterval(interval);
      resolve();
    }, RUN_DURATION_MS);
  });

  console.log(`\nStopping after ${beatCount} heartbeat(s)...`);

  // Step 4: Deregister and shutdown
  await client.deregisterAgent(instance_id);
  await client.shutdown();
  console.log('✓ Agent deregistered. Goodbye!');
}

main().catch((err: unknown) => {
  const message = err instanceof Error ? err.message : String(err);
  console.error('❌ Fatal error:', message);
  process.exit(1);
});

File: sdk/python/examples/auto_heartbeat.py
#!/usr/bin/env python3
"""
Example: Auto Heartbeat

This example shows how to run automatic heartbeat in background.
"""

import time
import threading
from eket_sdk import EketClient, AgentType, AgentRole, AgentStatus

# Initialize and register
client = EketClient(server_url="http://localhost:8080")
agent = client.register_agent(
    agent_type=AgentType.CUSTOM,
    role=AgentRole.SLAVER,
)

print(f"✅ Registered as {agent.instance_id}")


# Heartbeat thread
class HeartbeatThread(threading.Thread):
    def __init__(self, client: EketClient, interval: int = 60):
        super().__init__(daemon=True)
        self.client = client
        self.interval = interval
        self.running = True
        self.current_task = None
        self.current_status = AgentStatus.IDLE

    def run(self):
        """Send periodic heartbeats."""
        while self.running:
            try:
                response = self.client.send_heartbeat(
                    status=self.current_status, current_task=self.current_task
                )

                # Check for messages
                messages = response.get("messages", [])
                if messages:
                    print(f"📨 Received {len(messages)} message(s)")
                    for msg in messages:
                        print(f"   - {msg.get('type')}: {msg.get('payload')}")

                print(f"💓 Heartbeat sent (status: {self.current_status.value})")

            except Exception as e:
                print(f"❌ Heartbeat failed: {e}")

            time.sleep(self.interval)

    def stop(self):
        """Stop heartbeat thread."""
        self.running = False


# Start heartbeat
heartbeat_thread = HeartbeatThread(client, interval=30)
heartbeat_thread.start()

print("\n💓 Heartbeat started (every 30s)")
print("   Press Ctrl+C to stop\n")

# Simulate work
try:
    for i in range(10):
        time.sleep(5)

        # Simulate status changes
        if i % 3 == 0:
            heartbeat_thread.current_status = AgentStatus.IDLE
        else:
            heartbeat_thread.current_status = AgentStatus.ACTIVE

except KeyboardInterrupt:
    print("\n\n⏹️  Stopping...")

# Clean up
heartbeat_thread.stop()
client.deregister_agent()
client.close()

print("👋 Agent deregistered")
